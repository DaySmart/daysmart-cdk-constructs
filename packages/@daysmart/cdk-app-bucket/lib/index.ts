import cdk = require('@aws-cdk/core');
import s3 = require('@aws-cdk/aws-s3');

export class AppBucket extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: any) {
        super(scope, id);

        new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.stage}-${props.appName.toLowerCase()}-website-bucket`,
            publicReadAccess: true,
            versioned: true,
        });
    }
}