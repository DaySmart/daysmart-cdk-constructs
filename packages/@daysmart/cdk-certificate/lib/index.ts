import * as cdk from '@aws-cdk/core';
import acm = require('@aws-cdk/aws-certificatemanager');
import route53 = require('@aws-cdk/aws-route53');

export interface CdkCertificateProps {
    domainName: string;
    stage: string;
    project: string;
}

export class CdkCertificate extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkCertificateProps) {
    super(scope, id);

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: props.domainName
    });

    new acm.Certificate(this, 'Certificate', {
        domainName: props.domainName,
        subjectAlternativeNames: [
            `${props.stage}.${props.project}.${props.domainName}`,
            `*.${props.stage}.${props.project}.${props.domainName}`
        ],
        validation: acm.CertificateValidation.fromDns(hostedZone)
    });
  }
}
