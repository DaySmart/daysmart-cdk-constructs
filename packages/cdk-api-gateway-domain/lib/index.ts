import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';

export interface CdkApiGatewayDomainProps {
  companyDomainName: string;
  companyHostedZoneId: string;
  project: string;
  baseEnv: string;
  dynamicEnv?: string;
  certificateArn: string;
  restApiId: string;
  basePath: string;
}

export class CdkApiGatewayDomain extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkApiGatewayDomainProps) {
    super(scope, id);

    const api = apigw.RestApi.fromRestApiId(this, "Rest API", `${props.restApiId}`);

    let customDomain: apigw.DomainName;

    if (props.dynamicEnv) {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: `${props.dynamicEnv}-api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    } else {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: (props.baseEnv == "prod") ? `api.${props.project}.${props.companyDomainName}` : `api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    }

    const cloudformationBasePathMapping = new apigw.CfnBasePathMapping(this, "CloudformationBasePathMapping", {
      basePath: `${props.basePath}`,
      domainName: customDomain.domainName,
      restApiId: `${props.restApiId}`,
      stage: `${props.baseEnv}` 
    });

    const cloudformationRoute53Change = new route53.CfnRecordSet(this, "CloudformationAliasRecordRoute53", {
      aliasTarget: {
        dnsName: customDomain.domainNameAliasDomainName,
        evaluateTargetHealth: false,
        hostedZoneId: "Z2FDTNDATAQYW2"
      },
      comment: `CloudFront distribution for ${customDomain.domainName}`,
      hostedZoneId: props.companyHostedZoneId,
      name: customDomain.domainName,
      type: "A"
    });
  }
}
