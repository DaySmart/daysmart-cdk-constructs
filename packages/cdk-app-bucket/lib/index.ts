import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudfront as cloudfront } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

export interface AppBucketProps {
    stage: string;
    appName: string;
    dynamicEnvName: string;
    projectName: string;
    sharedServicesAccountId?: string;
}
export class AppBucket extends Construct {
    constructor(scope: Construct, id: string, props: AppBucketProps) {
        super(scope, id);

        const bucket = new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.projectName}`,
            versioned: true,
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL
        });

        const originAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'OriginAcessIdentity', {
            comment: `OriginAccessIdentity for ${bucket.bucketName}.`
        });

        bucket.grantRead(originAccessIdentity)

        if(props.sharedServicesAccountId) {
            bucket.grantReadWrite(new iam.AccountPrincipal(props.sharedServicesAccountId));
        }

        let output = new cdk.CfnOutput(this, "AppBucket", {
            value: bucket.bucketName
        });

        let output2 = new cdk.CfnOutput(this, "OriginAccessIdentity", {
            value: originAccessIdentity.originAccessIdentityName
        });

        output.overrideLogicalId("AppBucket");
        output2.overrideLogicalId("OriginAccessIdentity");
    }
}