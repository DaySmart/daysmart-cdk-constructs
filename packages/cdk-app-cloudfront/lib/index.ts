import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront'
import * as route53 from 'aws-cdk-lib/aws-route53'
import * as targets from 'aws-cdk-lib/aws-route53-targets'

export interface CdkAppCloudfrontProps {
    s3BucketName: string;
    originAccessIdentity: string;
    loggingBucketName?: string;
    stage: string;
    project: string;
    certArn: string;
    companyDomainName: string;
    companyHostedZoneId: string;
    dynamicEnvName?: string;
    appName?: string;
    projectHostedZoneId?: string;
    projectDomainName?: string;
}

export class CdkAppCloudfront extends Construct {

  constructor(scope: Construct, id: string, props: CdkAppCloudfrontProps) {
    super(scope, id);

    const codeBucket = s3.Bucket.fromBucketName(
        this,
        'CodeBucket',
        props.s3BucketName
    );

    const originAccessIdentity = cloudfront.OriginAccessIdentity.fromOriginAccessIdentityName(
        this,
        "OriginAccessIdentity",
        props.originAccessIdentity
    );

    let aliases: string[] = [];
    aliases.push(...this.getCompanyDomainAliases(props));

    if(props.projectDomainName && props.projectHostedZoneId) {
        aliases.push(...this.getProjectDomainAliases(props));
    }

    const distribution = new cloudfront.CloudFrontWebDistribution(this, 'Distribution', {
        originConfigs: [
            {
                s3OriginSource: {
                    s3BucketSource: codeBucket,
                    originAccessIdentity: originAccessIdentity
                },
                behaviors: [
                    {
                        isDefaultBehavior: true,
                        allowedMethods: cloudfront.CloudFrontAllowedMethods.ALL,
                        forwardedValues: {
                            queryString: true,
                            cookies: {
                              forward: "all",
                            },
                            headers: ["Referer"],
                        },
                    }
                ]
            }
        ],
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        priceClass: cloudfront.PriceClass.PRICE_CLASS_ALL,
        httpVersion: cloudfront.HttpVersion.HTTP2,
        loggingConfig: props.loggingBucketName ? {
            bucket: s3.Bucket.fromBucketName(
                this,
                "CloudfrontLoggingBucket",
                props.loggingBucketName
            ),
            prefix: `${props.stage}-${props.project}`,
            includeCookies: true
        } : undefined,
        errorConfigurations: [
            {
              errorCode: 404,
              responseCode: 200,
              errorCachingMinTtl: 1,
              responsePagePath: "/index.html",
            },
            {
              errorCode: 403,
              responseCode: 200,
              errorCachingMinTtl: 1,
              responsePagePath: "/index.html",
            },
        ],
        comment: `${props.stage} ${props.project}`,
        viewerCertificate: {
          props: {
            acmCertificateArn: props.certArn,
            sslSupportMethod: "sni-only",
            minimumProtocolVersion: "TLSv1.1_2016",
          },
          aliases: aliases
        },
    });

    const cloudfrontTarget = new targets.CloudFrontTarget(distribution);

    const companyHostedZone = route53.HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
        hostedZoneId: props.companyHostedZoneId,
        zoneName: props.companyDomainName
    });

    this.getCompanyDomainAliases(props).forEach(alias => {
        new route53.ARecord(this, alias, {
            zone: companyHostedZone,
            target: route53.RecordTarget.fromAlias(cloudfrontTarget),
            recordName: alias
        });
    });

    if(props.projectDomainName && props.projectHostedZoneId) {
        const projectHostedZone = route53.HostedZone.fromHostedZoneAttributes(this, 'ProjectHostedZone', {
            hostedZoneId: props.projectHostedZoneId,
            zoneName: props.projectDomainName
        });

        this.getProjectDomainAliases(props).forEach(alias => {
            new route53.ARecord(this, alias, {
                zone: projectHostedZone,
                target: route53.RecordTarget.fromAlias(cloudfrontTarget),
                recordName: alias
            });
        });
    }

    let output = new cdk.CfnOutput(this, "CloudfrontDistribution", {
        value: distribution.distributionId
    });

    let output2 = new cdk.CfnOutput(this, "DistributionDomainName", {
        value: distribution.distributionDomainName
    })

    output.overrideLogicalId("CloudfrontDistribution");
    output2.overrideLogicalId("DistributionDomainName");
  }

  getCompanyDomainAliases(props: CdkAppCloudfrontProps): Array<string> {
    let aliases: Array<string> = [];
    if(props.dynamicEnvName && props.appName) {
        aliases.push(`${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.project}.${props.companyDomainName}`);
    } else if(props.appName) {
        aliases.push(`${props.appName}.${props.stage}.${props.project}.${props.companyDomainName}`);
    } else {
        aliases.push(`${props.stage}.${props.project}.${props.companyDomainName}`);

        if(props.stage === 'prod') {
            aliases.push(`${props.project}.${props.companyDomainName}`);
        }
    }

    return aliases;
  }

  getProjectDomainAliases(props: CdkAppCloudfrontProps): Array<string> {
    let aliases: Array<string> = []
    if(props.dynamicEnvName && props.appName) {
        aliases.push(`${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.projectDomainName}`);
    } else if (props.appName) {
        aliases.push(`${props.appName}.${props.stage}.${props.projectDomainName}`);
    } else {
        aliases.push(`${props.stage}.${props.projectDomainName}`);
    }

    if(props.stage === "prod") {
        if(props.appName) {
            aliases.push(`${props.appName}.${props.projectDomainName}`);
        } else {
            aliases.push(props.projectDomainName as string);
        }
    }
    

    return aliases;
  }
}
