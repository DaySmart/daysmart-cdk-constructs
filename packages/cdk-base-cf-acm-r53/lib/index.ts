import * as cdk from '@aws-cdk/core';
import * as route53 from "@aws-cdk/aws-route53";
import * as targets from "@aws-cdk/aws-route53-targets";
import * as cloudfront from "@aws-cdk/aws-cloudfront";
import * as acm from "@aws-cdk/aws-certificatemanager";
import * as s3 from "@aws-cdk/aws-s3";

export interface CdkBaseCfAcmR53Props {
  defaultBehaviorOptions: cloudfront.BehaviorOptions;
  project: string;
  stage: string;
  loggingBucketName?: string;
  domains: string[];
}

export interface Zones {
  [key: string]: route53.IHostedZone
}

export class CdkBaseCfAcmR53 extends cdk.Construct {
  public distribution: cloudfront.Distribution;

  constructor(scope: cdk.Construct, id: string, props: CdkBaseCfAcmR53Props) {
    super(scope, id);

    let subjectAlternativeNames: Array<string> = [];
    let multiZones: Zones = {};
    const companyDomainNames = props.domains;

    companyDomainNames.forEach(companyDomainName => {
      var companyHostedZone = route53.HostedZone.fromLookup(this, `${companyDomainName} HostedZone`, {
        domainName: companyDomainName,
        privateZone: false
      });

      subjectAlternativeNames.push(
        `${props.stage}.${props.project}.${companyDomainName}`,
        `*.${props.stage}.${props.project}.${companyDomainName}`,
        `*.${companyDomainName}`
      );

      if (props.stage === 'prod') {
        subjectAlternativeNames.push(
          `${props.project}.${companyDomainName}`,
          `*.${props.project}.${companyDomainName}`
        );
      }

      multiZones[companyDomainName] = companyHostedZone
    });

    var certificateValidation = acm.CertificateValidation.fromDnsMultiZone(multiZones);

    var certificate = new acm.Certificate(this, 'Certificate', {
      domainName: subjectAlternativeNames[0],
      subjectAlternativeNames: subjectAlternativeNames.slice(1),
      validation: certificateValidation
    });

    this.distribution = new cloudfront.Distribution(this, "Distribution", {
      defaultBehavior: props.defaultBehaviorOptions,
      certificate: certificate,
      domainNames: this.getAliases(props, companyDomainNames),
      defaultRootObject: "index.html",
      comment: `${props.stage} ${props.project}`,
      enableLogging: (props.loggingBucketName) ? true : false,
      logBucket: (props.loggingBucketName) ? s3.Bucket.fromBucketName(this, "CloudfrontLoggingBucket", props.loggingBucketName) : undefined,
      logIncludesCookies: (props.loggingBucketName) ? true : false,
      logFilePrefix: (props.loggingBucketName) ? `${props.stage}-${props.project}` : undefined,
      httpVersion: cloudfront.HttpVersion.HTTP2,
      priceClass: cloudfront.PriceClass.PRICE_CLASS_ALL,
      errorResponses: [
        {
          httpStatus: 404,
          responseHttpStatus: 200,
          ttl: cdk.Duration.minutes(1),
          responsePagePath: "/index.html"
        },
        {
          httpStatus: 403,
          responseHttpStatus: 200,
          ttl: cdk.Duration.minutes(1),
          responsePagePath: "/index.html"
        },
      ],
      minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_1_2016
    });

    var cloudfrontTarget = new targets.CloudFrontTarget(this.distribution);

    companyDomainNames.forEach(companyDomainName => {
      var companyHostedZone = route53.HostedZone.fromLookup(this, `Same ${companyDomainName} HostedZone`, {
        domainName: companyDomainName,
        privateZone: false
      });
      new route53.ARecord(this, `${props.stage}.${props.project}.${companyDomainName}-RecordSet`, {
        zone: companyHostedZone,
        target: route53.RecordTarget.fromAlias(cloudfrontTarget),
        recordName: `${props.stage}.${props.project}.${companyDomainName}`
      });
      new route53.ARecord(this, `${props.stage}-${props.project}.${companyDomainName}-RecordSet`, {
        zone: companyHostedZone,
        target: route53.RecordTarget.fromAlias(cloudfrontTarget),
        recordName: `${props.stage}-${props.project}.${companyDomainName}`
      });
    });


    var output = new cdk.CfnOutput(this, "CertificateArn", {
      value: certificate.certificateArn
    });

    var output2 = new cdk.CfnOutput(this, "CloudfrontDistribution", {
      value: this.distribution.distributionId
    });

    var output3 = new cdk.CfnOutput(this, "DistributionDomainName", {
      value: this.distribution.distributionDomainName
    })

    output.overrideLogicalId("CertificateArn");
    output2.overrideLogicalId("CloudfrontDistribution");
    output3.overrideLogicalId("DistributionDomainName");
  }

  getAliases(props: CdkBaseCfAcmR53Props, companyDomainNames: Array<string>): Array<string> {
    let aliases: Array<string> = [];
    companyDomainNames.forEach(companyDomainName => {
      aliases.push(
        `${props.stage}.${props.project}.${companyDomainName}`,
        `${props.stage}-${props.project}.${companyDomainName}`
      );
    });
    
    return aliases;
  }
}
