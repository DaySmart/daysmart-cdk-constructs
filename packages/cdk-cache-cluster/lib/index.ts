import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_elasticache as elasticache } from 'aws-cdk-lib';

export interface CdkCacheClusterProps {
    cacheNodeType: string,
    engine: string,
    numCacheNodes: number,
    clusterName?: string
}

export class CdkCacheCluster extends Construct {
    constructor(scope: Construct, id: string, props: CdkCacheClusterProps) {
      super(scope, id);

      const cluster = new elasticache.CfnCacheCluster(this, 'MyCluster', {
        cacheNodeType: props.cacheNodeType,
        engine: props.engine,
        numCacheNodes: props.numCacheNodes
      })
    }
}