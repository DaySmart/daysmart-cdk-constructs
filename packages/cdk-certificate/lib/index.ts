import * as cdk from '@aws-cdk/core';
import acm = require('@aws-cdk/aws-certificatemanager');
import route53 = require('@aws-cdk/aws-route53');

export interface CdkCertificateProps {
    domainName: string;
    stage: string;
    project: string;
    hostedZoneId: string;
}

export class CdkCertificate extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkCertificateProps) {
    super(scope, id);

    const hostedZone = route53.HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
        hostedZoneId: props.hostedZoneId,
        zoneName: props.domainName
    });

    const cert = new acm.Certificate(this, 'Certificate', {
        domainName: props.domainName,
        subjectAlternativeNames: [
            `${props.stage}.${props.project}.daysmart.com`,
            `*.${props.stage}.${props.project}.daysmart.com`
        ],
        validation: acm.CertificateValidation.fromDns(hostedZone)
    });

    let output = new cdk.CfnOutput(this, "CertificateArn",{
        value: cert.certificateArn
    });

    output.overrideLogicalId("CertificateArn");
  }
}
