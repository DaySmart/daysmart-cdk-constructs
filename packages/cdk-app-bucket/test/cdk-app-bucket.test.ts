import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { AppBucket } from '../lib/index';

test("AppBucket", () => {
    const stack = new Stack();
    new AppBucket(stack, 'AppBucket', {
        stage: 'test',
        appName: 'test2',
        dynamicEnvName: 'test3',
        projectName: 'test4',
        sharedServicesAccountId: '1234567'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
})
