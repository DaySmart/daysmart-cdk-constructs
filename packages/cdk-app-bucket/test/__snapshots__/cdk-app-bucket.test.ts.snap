// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBucket 1`] = `
Object {
  "Outputs": Object {
    "AppBucket": Object {
      "Value": Object {
        "Ref": "AppBucketFED55033",
      },
    },
    "OriginAccessIdentity": Object {
      "Value": Object {
        "Ref": "AppBucketOriginAcessIdentityEAF14A7F",
      },
    },
  },
  "Resources": Object {
    "AppBucketFED55033": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "test3-test2.test.test4",
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "AppBucketOriginAcessIdentityEAF14A7F": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "OriginAccessIdentity for ",
                Object {
                  "Ref": "AppBucketFED55033",
                },
                ".",
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "AppBucketPolicyF441A69E": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "AppBucketFED55033",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "AppBucketOriginAcessIdentityEAF14A7F",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "AppBucketFED55033",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "AppBucketFED55033",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::1234567:root",
                    ],
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "AppBucketFED55033",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "AppBucketFED55033",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
